LLVM: add built-in lld linker support for x86.

chen-png opened the pull request:
add support to use LLVM built-in lld linker to build zephyr
for x86 platforms, mainly include:

add a CONFIG_LLVM_USE_LD to select the default linker for LLVM, ld or lld.
remove unsupported --sort-common option for lld; 
add linker command to handle .symtab, .strtab and .shstrtab sections for lld, because they will still be treated as a warning even with --orphan-handling=warn;
discard .eh_frame section if without CONFIG_EXCEPTIONS.
keep those sections related to ISR(.text.irqstubs, .gnu.linkonce.t.exc*) in output section for lld, because they won't be linked into output due to symbols unreferenced, and there are some difference of the gc-sections behavior between ld and lld.


fix:  #41776 

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

chen-png commented:
tested on qemu_x86, most of testcases worked well except for 6 failures, will try to fix them.
command: twister -W -p qemu_x86 -T tests/kernel/ -x=CONFIG_LLVM_USE_BUILTIN_LLD=y 
result:
>INFO    -  76/101 qemu_x86                  tests/kernel/threads/thread_init/kernel.threads.init FAILED Exited with 2 (qemu)
ERROR   - see: /home/chenpeng/zephyrproject/zephyr/twister-out/qemu_x86/tests/kernel/threads/thread_init/kernel.threads.init/handler.log
INFO    -  77/101 qemu_x86                  tests/kernel/mutex/sys_mutex/system.mutex.nouser   PASSED (qemu 6.435s)
INFO    -  78/101 qemu_x86                  tests/kernel/threads/thread_error_case/kernel.threads.error.case PASSED (qemu 6.656s)
INFO    -  79/101 qemu_x86                  tests/kernel/threads/thread_stack/kernel.threads.thread_stack PASSED (qemu 7.242s)
INFO    -  80/101 qemu_x86                  tests/kernel/mutex/sys_mutex/system.mutex          PASSED (qemu 6.552s)
INFO    -  81/101 qemu_x86                  tests/kernel/mutex/mutex_error_case/kernel.mutex_error_case PASSED (qemu 6.633s)
INFO    -  82/101 qemu_x86                  tests/kernel/mutex/mutex_api/kernel.mutex          PASSED (qemu 6.711s)
INFO    -  83/101 qemu_x86                  tests/kernel/mem_slab/mslab_concept/kernel.memory_slabs.concept PASSED (qemu 6.511s)
INFO    -  84/101 qemu_x86                  tests/kernel/mem_slab/mslab_api/kernel.memory_slabs.api PASSED (qemu 6.367s)
INFO    -  85/101 qemu_x86                  tests/kernel/mem_slab/mslab_threadsafe/kernel.memory_slabs.threadsafe PASSED (qemu 6.521s)
INFO    -  86/101 qemu_x86                  tests/kernel/mem_slab/mslab/kernel.memory_slabs.api PASSED (qemu 6.496s)
INFO    -  87/101 qemu_x86                  tests/kernel/workq/critical/kernel.workqueue.sam   SKIPPED (filter)
INFO    -  88/101 qemu_x86                  tests/kernel/tickless/tickless_concept/kernel.tickless.concept PASSED (qemu 8.435s)
INFO    -  89/101 qemu_x86                  tests/kernel/profiling/profiling_api/kernel.common.profiling PASSED (qemu 6.469s)
INFO    -  90/101 qemu_x86                  tests/kernel/semaphore/sys_sem/kernel.semaphore.usage PASSED (qemu 6.498s)
INFO    -  91/101 qemu_x86                  tests/kernel/semaphore/semaphore/kernel.semaphore  PASSED (qemu 6.795s)
INFO    -  92/101 qemu_x86                  tests/kernel/workq/work_queue/kernel.workqueue     PASSED (qemu 6.600s)
INFO    -  93/101 qemu_x86                  tests/kernel/queue/kernel.queue                    PASSED (qemu 6.845s)
INFO    -  94/101 qemu_x86                  tests/kernel/workq/critical/kernel.workqueue       PASSED (qemu 6.457s)
INFO    -  95/101 qemu_x86                  tests/kernel/workq/work/kernel.work.api            PASSED (qemu 6.388s)
INFO    -  96/101 qemu_x86                  tests/kernel/pending/kernel.objects                PASSED (qemu 5.498s)
INFO    -  97/101 qemu_x86                  tests/kernel/workq/user_work/kernel.work.user      PASSED (qemu 5.132s)
INFO    -  98/101 qemu_x86                  tests/kernel/stack/stack/kernel.stack.usage        PASSED (qemu 4.361s)
INFO    -  99/101 qemu_x86                  tests/kernel/device/kernel.device.pm               PASSED (qemu 4.430s)
INFO    - 100/101 qemu_x86                  tests/kernel/device/kernel.device                  PASSED (qemu 4.604s)
INFO    - 101/101 qemu_x86                  tests/kernel/poll/kernel.poll                      PASSED (qemu 4.701s)
INFO    - **78 of 84 test configurations passed (92.86%), 6 failed, 83 skipped with 0 warnings in 402.80 seconds**
INFO    - In total 958 test cases were executed, 673 skipped on 1 out of total 418 platforms (0.24%)
INFO    - 77 test configurations executed on platforms, 7 test configurations were only built.

chen-png commented:
some sections, such as ".text.irqstubs" , ".gnu.linkonce.t.*", they will be discarded by lld linker, but we need them to get all interrupt vector information, so add KEEP command explicitly to put into output.
but if we did this, we might meet the error ""Multiple assignments for vector ".

i.e. in arch/x86/core/ia32/float.c, " _FpNotAvailableExcHandler" is defined to handle exception vector 7,
but in arch/x86/core/ia32/fatal.c, we also defined a handle for vector 7 "EXC_FUNC_NOCODE(IV_DEVICE_NOT_AVAILABLE, 0);",
then it reported a error "Multiple assignments for vector 7" by arch/x86/gen_idt.py.

chen-png commented:
@tejlmand Could you help take a look of this? thanks.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -162,6 +162,12 @@ config CMAKE_LINKER_GENERATOR
 
 endchoice
 
+config LLVM_USE_BUILTIN_LLD
+	bool "LLVM uses built-in lld linker"
```
Maybe we should consider `lld` to be the linker for `llvm`, and then have a setting requesting `ld` to be used as linker.

We can still default `llvm` to use `ld`, as we do today.

For example:
```
config LLVM_USE_LD
	bool "LLVM uses ld linker"
	default y
	help
	  Use binutils ld linker instead of LLVM built-in lld linker.
```

chen-png replied:
thanks, have updated it using your suggestions.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -162,6 +162,12 @@ config CMAKE_LINKER_GENERATOR
 
+config LLVM_USE_BUILTIN_LLD
+	bool "LLVM uses built-in lld linker"
+	default n
```
We should have this setting depending on llvm being the toolchain in use.

tejlmand commented on include/arch/x86/ia32/linker.ld:
```diff
@@ -308,11 +308,17 @@ SECTIONS
  * zephyr_linker_sources(ROM_START ...). This typically contains the vector
  * table and debug information.
  */
+#ifndef CONFIG_LLVM_USE_BUILTIN_LLD
```
why should those user-snippets not be included when using lld ?

chen-png replied:
because in user-snippets, it will set local counter to 0x0, this will cause a move afterward error for those platforms whose rom start offset is not 0x0, but this could be fixed by setting CONFIG_ROM_START_OFFSET according to your suggestions, so remove this from code change.

tejlmand commented on include/arch/x86/ia32/linker.ld:
```diff
@@ -533,6 +539,13 @@ SECTIONS
 	/DISCARD/ : { *(.note.GNU-stack) }
 
+#ifdef CONFIG_LLVM_USE_BUILTIN_LLD
+	/DISCARD/ : { *(.eh_frame) }
```
why do we need to discard the `.eh_frame` for lld ?

If C++ code is compiled with exceptions, then I believe the `.eh_frame` section should be kept.
Else exceptions should probably be disabled with `CONFIG_EXCEPTIONS=n`
https://github.com/zephyrproject-rtos/zephyr/blob/280309e636cb728339b01b52d12b7e2f52216fc7/CMakeLists.txt#L232-L234

chen-png replied:
even without CONFIG_EXCEPTIONS, gun ld and llvm lld they all will generate eh_frame section, but for ld, it will be removed using "-gc-sections" as it is unused, however, for lld, it is still saved even with "-gc-sections", if we don't handle it in linker script, it will report a warning.
so discard it explicitly, because if without CONFIG_EXCEPTIONS, we don't need this section actually, if with CONFIG_EXCEPTIONS, it will be handled by cplusplus-rom.ld before.

tejlmand reviewed (dismissed):
thanks for taking Zephyr one step closed to support `lld`.

Some observations.

carlescufi commented:
@chen-png could you please address the review comments?

chen-png commented:
>could you please address the review comments?

so sorry for the late response, will address them soon.

chen-png commented:
the "--gc-sections" behavior between gnu ld and llvm lld  is a little different, even with the same input object file, the removed unused sections are different, i.e.
lld removed some necessary sections ".text.irqstubs", ".text.hpet_isr", we need to use them for extracting interrupt information.
>removed sections in hpet.c.obj by gnu ld
```
/usr/bin/ld: removing unused section '.text.smp_timer_init' in file 'zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj)'
/usr/bin/ld: removing unused section '.text.sys_clock_cycle_get_32' in file 'zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj)'
/usr/bin/ld: removing unused section '.text.sys_clock_cycle_get_64' in file 'zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj)'
/usr/bin/ld: removing unused section '.text.sys_clock_idle_exit' in file 'zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj)'
```
>removed sections in hpet.c.obj by llvm lld
```
removing unused section zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj):(.text)
removing unused section zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj):(.text.smp_timer_init)
removing unused section zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj):(.text.sys_clock_cycle_get_32)
removing unused section zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj):(.text.sys_clock_cycle_get_64)
removing unused section zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj):(.text.sys_clock_idle_exit)
removing unused section zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj):(.text.irqstubs)
removing unused section zephyr/drivers/timer/libdrivers__timer.a(hpet.c.obj):(.text.hpet_isr)
```

Commit:
LLVM: add built-in lld linker support for x86.

add support to use LLVM built-in lld linker to build
zephyr applications for x86 platforms.

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

chen-png commented on cmake/linker/ld/target_base.cmake:
```diff
@@ -24,8 +24,15 @@ macro(toolchain_ld_base)
+      CONFIG_LINKER_SORT_BY_ALIGNMENT
+      ${LINKERFLAGPREFIX},--sort-common=descending
+    )
+  endif()
```
added a specific target_base.cmake for lld.

chen-png commented on include/linker/linker-tool.h:
```diff
@@ -17,6 +17,13 @@
 #if defined(_LINKER)
 #if defined(__GCC_LINKER_CMD__)
+/* ALIGN_WITH_INPUT is not supported by LLVM lld,
+ * it needs to be replaced by ALIGN(8).
```
for lld, it seems that ALIGN_WITH_INPUT is the default alignment, see discussions in https://reviews.llvm.org/D74286.

chen-png commented:
hi @tejlmand Could you help take a look of this? thanks, sorry for the late response.

chen-png commented:
hi @tejlmand @dcpleung @jhedberg Could you help take a look of this again? thanks.

chen-png commented:
> @chen-png can you please rebase this?

yes, done.
